crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,266 @@
+        Ok(Plugin(Arc::new(Repr {
+            store: Mutex::new(store),
+            memory,
+            free_function,
+            functions,
+            bytes: bytes.as_ref().into(),
+        })))
+    }
+
+    fn store(&self) -> MutexGuard<'_, Store> {
+        self.0.store.lock().unwrap()
+    }
+
+    /// Call a function defined in the plugin under `function_name`.
+    ///
+    /// This will eat the number of argument it needs of type Bytes.
+    ///
+    /// # Errors
+    /// - if the plugin doesn't contain the function
+    /// - if the number of argument isn't correct
+    pub fn call(&self, function_name: &str, args: &mut Args) -> SourceResult<Value> {
+        let span = args.span;
+        let ty = self
+            .get_function(function_name)
+            .ok_or("plugin doesn't have the method: {function}")
+            .at(span)?
+            .ty(self.store().as_context());
+        let arg_count = ty.params().len();
+        let mut byte_args = vec![];
+        for k in 0..arg_count {
+            let arg = args
+                .eat::<Value>()?
+                .ok_or(format!("plugin methods takes {arg_count} args, {k} provided"))
+                .at(span)?
+                .cast::<Bytes>()
+                .at(span)?;
+            byte_args.push(arg);
+        }
+        let byte_args = byte_args.iter().map(|b| b.as_slice()).collect::<Vec<_>>();
+        let s = self.call_inner(function_name, byte_args).at(span)?;
+        Ok(Value::Bytes(s.get().into()))
+    }
+
+    fn call_inner<'a>(
+        &self,
+        function: &str,
+        args: impl IntoIterator<Item = &'a [u8]>,
+    ) -> StrResult<ReturnedData> {
+        self.store().data_mut().result_ptr = 0;
+        self.store().data_mut().result_len = 0;
+
+        let mut result_args = Vec::new();
+        let mut guard = self.store();
+        let arg_buffer = &mut guard.data_mut().arg_buffer;
+        arg_buffer.clear();
+        for arg in args {
+            result_args.push(WasiValue::I32(arg.len() as _));
+            arg_buffer.extend_from_slice(arg);
+        }
+        drop(guard);
+
+        let function = self
+            .get_function(function)
+            .ok_or(format!("plugin doesn't have the method: {function}"))?;
+
+        let mut code = WasiValue::I32(2);
+        let ty = function.ty(self.store().as_context());
+        if ty.params().len() != result_args.len() {
+            return Err("incorrect number of arguments".into());
+        }
+
+        let call_result = function.call(
+            &mut self.store().as_context_mut(),
+            &result_args,
+            std::array::from_mut(&mut code),
+        );
+        let (ptr, len) = (self.store().data().result_ptr, self.store().data().result_len);
+        let result = ReturnedData {
+            memory: self.0.memory,
+            ptr,
+            len,
+            free_function: &self.0.free_function,
+            context_mut: self.store(),
+        };
```

Comment:
Ok I think got why some tests deadlock: one of those holds a lock to the store, preventing further `self.store()` calls from completing.

The fix should be simple: put the store in its own variable, and use the variable:
```rust
let store = self.store();
store.data_mut().result_ptr = 0;
store.data_mut().result_len = 0;
// ...
let arg_buffer = &mut store.data_mut().arg_buffer;
// etc
```

Reply:
I don't think so since the test containing only one call doesn't freeze?

Reply:
Really ? That's weird, as making this change made all the tests pass for me (before the protocol was reverted)

Anyway, even with the old protocol, I think this might be a valuable change to make, as it avoids a lot of calls to `self.store()`, and thus a lot of locking/unlocking of the mutex.